api/src/main/java/org/projectnessie/tools/catalog/migration/api/CatalogMigrator.java

```diff
@@ -0,0 +1,247 @@
+/*
+ * Copyright (C) 2023 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.tools.catalog.migration.api;
+
+import com.google.common.base.Preconditions;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.iceberg.BaseTable;
+import org.apache.iceberg.TableOperations;
+import org.apache.iceberg.catalog.Catalog;
+import org.apache.iceberg.catalog.Namespace;
+import org.apache.iceberg.catalog.SupportsNamespaces;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.exceptions.AlreadyExistsException;
+import org.apache.iceberg.exceptions.NoSuchNamespaceException;
+import org.apache.iceberg.hadoop.HadoopCatalog;
+import org.immutables.value.Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Value.Immutable
+public abstract class CatalogMigrator {
+
+  /** Source {@link Catalog} from which the tables are chosen. */
+  public abstract Catalog sourceCatalog();
+
+  /** Target {@link Catalog} to which the tables need to be registered or migrated. */
+  public abstract Catalog targetCatalog();
+
+  /** Delete the table entries from the source catalog after successful registration. */
+  public abstract boolean deleteEntriesFromSourceCatalog();
+
+  /** Enable the stacktrace in logs in case of failures. */
+  @Value.Default
+  public boolean enableStacktrace() {
+    return false;
+  }
+
+  @Value.Check
+  void check() {
+    if (!(targetCatalog() instanceof SupportsNamespaces)) {
+      throw new UnsupportedOperationException(
+          String.format(
+              "target catalog %s doesn't implement SupportsNamespaces to create missing namespaces.",
+              targetCatalog().name()));
+    }
+
+    if (!(sourceCatalog() instanceof SupportsNamespaces)) {
```

Comment:
I am surprised that this condition now appears here.
Does this mean that there are no catalogs that do not implement `SupportsNamespaces` and all the work and back-and-forth was just wasted effort?

Reply:
All the iceberg catalogs that are supported by this tool (maybe excluding custom catalog) implement `SupportsNamespaces`. There can be a custom catalog without `SupportsNamespaces` but I don't want to drag this discussion based on an imaginary catalog. 

If such catlog exists in future, the user can raise a requirement and we can support it.